From: Robert Millan <rmh@debian.org>
Date: Sat, 21 Dec 2013 17:04:48 -0700
Subject: Report correct disk size on GNU/kFreeBSD. Thanks Tuco.

Addresses-Debian-Bug: 592292
---
 lib/blkdev.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/blkdev.c b/lib/blkdev.c
index f8182c0..a293529 100644
--- a/lib/blkdev.c
+++ b/lib/blkdev.c
@@ -25,6 +25,10 @@
 #include <sys/disk.h>
 #endif
 
+#ifdef __FreeBSD_kernel__
+#include <sys/disk.h>
+#endif
+
 #include "blkdev.h"
 #include "c.h"
 #include "linux_version.h"
